src/query/models/tags.go

```diff
@@ -93,37 +88,118 @@ func (t Tags) IDLen() int {
 	return idLen
 }
 
-// IDWithExcludes returns a string representation of the tags excluding some tag keys.
-func (t Tags) IDWithExcludes(excludeKeys ...[]byte) uint64 {
-	b := make([]byte, 0, t.Len())
-	for _, tag := range t.Tags {
-		// Always exclude the metric name by default
-		if bytes.Equal(tag.Name, t.Opts.MetricName()) {
-			continue
+func (t Tags) quotedID() []byte {
+	needEscaping, length := t.escapingAndLength()
+	if needEscaping == nil {
+		return t.quoteIDSimple(length)
+	}
+
+	// TODO: pool these bytes.
+	id := make([]byte, length)
+	idx := 0
+	for i, tag := range t.Tags {
+		if needEscaping[i*2] {
+			idx = strconv.Escape(id, tag.Name, idx)
+		} else {
+			idx += copy(id[idx:], tag.Name)
 		}
 
-		found := false
-		for _, n := range excludeKeys {
-			if bytes.Equal(n, tag.Name) {
-				found = true
-				break
+		if needEscaping[i*2+1] {
+			idx = strconv.Quote(id, tag.Value, idx)
+		} else {
+			idx = strconv.QuoteSimple(id, tag.Value, idx)
+		}
+	}
+
+	return id
+}
+
+// adds quotes to tag values when no characters need escaping.
+func (t Tags) quoteIDSimple(length int) []byte {
+	// TODO: pool these bytes.
+	id := make([]byte, length)
+	idx := 0
+	for _, tag := range t.Tags {
+		idx += copy(id[idx:], tag.Name)
+		idx = strconv.QuoteSimple(id, tag.Value, idx)
+	}
+
+	return id
+}
+
+func (t Tags) escapingAndLength() ([]bool, int) {
```

Comment:
One potential way to make this signature more clear without compromising perf (I think): 

```
type tagEscaping struct {
    EscapeName bool
    EscapeValue bool
}

func (t Tags) escapingAndLength() ([]tagEscaping, int) {
...
```
https://play.golang.org/p/ctnDxYvBxyK shows that the size of the two is the same (i.e. an array of 2 of the struct takes up the same space as 4 bools)

Reply:
Cool makes sense. Actually now that I think about it, what if we just moved the `Quote` vs simple quote decision and actual quoting into a single method? That is, make a method which: 

1. Checks the string for characters needing escaping + calculate the length
2. Allocates the appropriate byte array.
3. Calls `Quote` or `QuoteSimple` as appropriate

Perf wise I think it would be equivalent, and it would abstract away some of the details from `Tags`, allowing this class to just call `Quote`, which does the right thing and is performant. 

Thoughts?

Reply:
Unfortunately we need to calculate the length for the whole tag set, otherwise we're allocating a bunch of temporary buffers per-tag, then putting them together. If `Tags` handles this itself, it can determine the total buffer size for the whole ID